### PR TITLE
planner: donot prune all columns for Projection

### LINKED ISSUE #23887: executor: (*Column).GetString panic
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
mysql -h172.16.5.85 -P52324 -uroot -D subquery_agg10000

```
SELECT  ( 3, 5 )  IN ( SELECT   SUBQUERY1_t1 . `col_decimal_30_10_not_null` AS SUBQUERY1_field1 , SUBQUERY1_t2 . `col_int_not_null_key` AS SUBQUERY1_field2 FROM `V` AS SUBQUERY1_t1 RIGHT  JOIN `H` AS SUBQUERY1_t2 ON (SUBQUERY1_t2 . `col_varchar_key` = SUBQUERY1_t1 . `col_varchar_64_not_null`  ) WHERE ( SUBQUERY1_t1 . `col_varchar_64` <= 'y' AND SUBQUERY1_t1 . `col_char_64` <> 't' ) GROUP BY SUBQUERY1_field1 , SUBQUERY1_field2  ) AS field1 FROM `J` AS table1 LEFT  JOIN `W` AS table2 ON (table2 . `col_decimal_40_not_null_key` = table1 . `col_decimal_30_10_key`  ) WHERE  ( 'f', 'e' ) NOT IN ( SELECT   SUBQUERY2_t1 . `col_varchar_64` AS SUBQUERY2_field1 , min( SUBQUERY2_t1 . `col_char_64` ) AS SUBQUERY2_field2 FROM `Z` AS SUBQUERY2_t1 LEFT OUTER JOIN `B` AS SUBQUERY2_t2 ON (SUBQUERY2_t2 . `col_char_64` = SUBQUERY2_t1 . `col_varchar_key`  )  GROUP BY SUBQUERY2_field1  ) AND ( table2 . `col_char_64` != 'w' OR table1 . `col_int_not_null` IS  NULL )   ;
```
```
SELECT  ( 'l', 'f' )  IN ( SELECT DISTINCT  SUBQUERY1_t2 . `col_varchar_64_not_null` AS SUBQUERY1_field1 , max( SUBQUERY1_t1 . `col_varchar_64` ) AS SUBQUERY1_field2 FROM `T` AS SUBQUERY1_t1 LEFT OUTER JOIN `O` AS SUBQUERY1_t2 ON (SUBQUERY1_t2 . `col_varchar_64` = SUBQUERY1_t1 . `col_varchar_key`  )  GROUP BY SUBQUERY1_field1  ) AS field1 FROM `DD` AS table1 RIGHT OUTER JOIN `X` AS table2 INNER JOIN `P` AS table3 ON (table3 . `col_int_key` = table2 . `col_decimal_not_null_key`  ) ON (table3 . `col_decimal` = table2 . `col_decimal_40_key`  ) WHERE  table1 . `col_char_64_not_null` > ANY ( SELECT   SUBQUERY2_t1 . `col_char_64_not_null` AS SUBQUERY2_field1 FROM `I` AS SUBQUERY2_t1 RIGHT  JOIN `J` AS SUBQUERY2_t2 ON (SUBQUERY2_t2 . `col_decimal_40_key` = SUBQUERY2_t1 . `col_tinyint_not_null`  ) WHERE SUBQUERY2_t1 . `col_int_not_null` <> table3 . `col_decimal_40` AND SUBQUERY2_t1 . `col_int` != 6 ) OR table1 . `pk` = 224 OR (  ( SELECT   COUNT(  SUBQUERY3_t1 . `col_varchar_64` ) AS SUBQUERY3_field1 FROM `B` AS SUBQUERY3_t1 INNER JOIN `W` AS SUBQUERY3_t2 ON (SUBQUERY3_t2 . `pk` = SUBQUERY3_t1 . `col_decimal_key`  )  ) IS  NULL AND table1 . `col_int` >= table2 . `col_decimal_key` )   ;
ERROR 1105 (HY000): runtime error: index out of range [0] with length 0
```
```
ERROR 1105 (HY000) at line 1: runtime error: index out of range [9] with length 9
SELECT  ( 8, 2 )  IN ( SELECT   SUBQUERY1_t1 . `col_decimal_30_10` AS SUBQUERY1_field1 , max( SUBQUERY1_t1 . `col_decimal_40_key` ) AS SUBQUERY1_field2 FROM `O` AS SUBQUERY1_t1 RIGHT OUTER JOIN ( `U` AS SUBQUERY1_t2 RIGHT  JOIN `X` AS SUBQUERY1_t3 ON (SUBQUERY1_t3 . `col_varchar_key` = SUBQUERY1_t2 . `col_varchar_64_not_null`  )) ON (SUBQUERY1_t3 . `col_char_64` = SUBQUERY1_t2 . `col_varchar_key`  )  GROUP BY SUBQUERY1_field1  ) AS field1 FROM `G` AS table1 LEFT OUTER JOIN `CC` AS table2 ON (table2 . `col_varchar_64_not_null` = table1 . `col_varchar_key`  ) WHERE  NOT EXISTS ( SELECT   SUBQUERY2_t1 . `col_varchar_64` AS SUBQUERY2_field1 FROM `H` AS SUBQUERY2_t1 LEFT  JOIN `F` AS SUBQUERY2_t2 ON (SUBQUERY2_t2 . `col_bigint` = SUBQUERY2_t1 . `col_decimal_40_key`  ) WHERE ( SUBQUERY2_t1 . `col_decimal_40` <> 9 AND SUBQUERY2_t2 . `col_varchar_64_not_null` = 'o' ) GROUP BY SUBQUERY2_field1 ) AND table2 . `col_char_64` >= 't'   ;
```

### 2. What did you expect to see? (Required)
ok

### 3. What did you see instead (Required)
```
2021/04/07 15:16:23.293 +08:00] [ERROR] [misc.go:90] ["panic in the recoverable goroutine"] [r={}] ["stack trace"="github.com/pingcap/tidb/util.WithRecovery.func1\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/util/misc.go:92\nruntime.gopanic\n\t/usr/local/go/src/runtime/panic.go:679\nruntime.goPanicIndex\n\t/usr/local/go/src/runtime/panic.go:75\ngithub.com/pingcap/tidb/util/chunk.(*Column).GetString\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/util/chunk/column.go:528\ngithub.com/pingcap/tidb/util/chunk.Row.GetString\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/util/chunk/row.go:72\ngithub.com/pingcap/tidb/expression.(*Column).EvalString\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/expression/column.go:397\ngithub.com/pingcap/tidb/expression.CompareStringWithCollationInfo\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/expression/builtin_compare.go:2657\ngithub.com/pingcap/tidb/expression.(*builtinEQStringSig).evalInt\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/expression/builtin_compare.go:2135\ngithub.com/pingcap/tidb/expression.(*ScalarFunction).EvalInt\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/expression/scalar_function.go:378\ngithub.com/pingcap/tidb/expression.(*ScalarFunction).Eval\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/expression/scalar_function.go:345\ngithub.com/pingcap/tidb/expression.EvalBool\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/expression/expression.go:235\ngithub.com/pingcap/tidb/executor.(*antiSemiJoiner).tryToMatchInners\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/executor/joiner.go:449\ngithub.com/pingcap/tidb/executor.(*HashJoinExec).joinMatchedProbeSideRow2Chunk\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/executor/join.go:527\ngithub.com/pingcap/tidb/executor.(*HashJoinExec).join2Chunk\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/executor/join.go:596\ngithub.com/pingcap/tidb/executor.(*HashJoinExec).runJoinWorker\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/executor/join.go:457\ngithub.com/pingcap/tidb/executor.(*HashJoinExec).fetchAndProbeHashTable.func2\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/executor/join.go:339\ngithub.com/pingcap/tidb/util.WithRecovery\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/util/misc.go:95"]
```
```
[ERROR] [misc.go:90] ["panic in the recoverable goroutine"] [r={}] ["stack trace"="github.com/pingcap/tidb/util.WithRecovery.func1\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/util/misc.go:92\nruntime.gopanic\n\t/usr/local/go/src/runtime/panic.go:679\nruntime.goPanicIndex\n\t/usr/local/go/src/runtime/panic.go:75\ngithub.com/pingcap/tidb/util/chunk.(*Column).GetString\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/util/chunk/column.go:528\ngithub.com/pingcap/tidb/util/chunk.Row.GetString\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/util/chunk/row.go:72\ngithub.com/pingcap/tidb/expression.(*Column).EvalString\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/expression/column.go:397\ngithub.com/pingcap/tidb/expression.CompareStringWithCollationInfo\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/expression/builtin_compare.go:2657\ngithub.com/pingcap/tidb/expression.(*builtinEQStringSig).evalInt\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/expression/builtin_compare.go:2135\ngithub.com/pingcap/tidb/expression.(*ScalarFunction).EvalInt\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/expression/scalar_function.go:378\ngithub.com/pingcap/tidb/expression.(*ScalarFunction).Eval\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/expression/scalar_function.go:345\ngithub.com/pingcap/tidb/expression.EvalBool\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/expression/expression.go:235\ngithub.com/pingcap/tidb/executor.(*leftOuterSemiJoiner).tryToMatchInners\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/executor/joiner.go:519\ngithub.com/pingcap/tidb/executor.(*HashJoinExec).joinMatchedProbeSideRow2Chunk\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/executor/join.go:527\ngithub.com/pingcap/tidb/executor.(*HashJoinExec).join2Chunk\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/executor/join.go:596\ngithub.com/pingcap/tidb/executor.(*HashJoinExec).runJoinWorker\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/executor/join.go:457\ngithub.com/pingcap/tidb/executor.(*HashJoinExec).fetchAndProbeHashTable.func2\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/executor/join.go:339\ngithub.com/pingcap/tidb/util.WithRecovery\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/util/misc.go:95"]
```
### 4. What is your TiDB version? (Required)
master of nightly

### PR DESCRIPTION
**Pull request by guo-shaoge**
Signed-off-by: guo-shaoge <shaoge1994@163.com>

### What problem does this PR solve?

Issue Number: close #23887 

Problem Summary:  

```
Hash Join (eq("abc", tbl1.col1))
  |                     |
Projection     DataSource(tbl1)
  |
Index Join
```

1. We prune all columns in `Projection`
2. Resolve indices for `col1`, Its index is 0(because `Projection`'s output columns are all pruned)
3. `Projection` is eliminated, because its output columns are all pruned. But **Index Join** still has output columns
4. Got panic when calc `eq("abc", tbl1.col1)`. Because use index 0 to eval tbl1.col1

### What is changed and how it works?
What's Changed: 
1. We donot prune all columns in `Projection`. If `Projection` is pruned to empty, we make it a `Projection 1` (just like count(1))
2. Delete some code in `rule_eliminate_projection.go`, because Projection cannot be empty. Check [here](https://github.com/pingcap/tidb/pull/6318)

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Integration test

Side effects: no

### Release note 

- donot prune all columns for Projection

### CODE REVIEW

**Comment by guo-shaoge** [comment]
@winoros @lzmhhh123 PTAL

**Comment by guo-shaoge** [comment]
As @lzmhhh123 suggested, I'm trying to make `Projection(1)`, just like `Count(1)` does. Good for performance.

**Review by lzmhhh123 — COMMENTED** [review]
How about adding some tests? BTW, can we avoid pruning projection in `tryEliminateProjection`?

**Comment by lzmhhh123** [comment]
@winoros PTAL the PR. Do you have any better way to solve the issue?

**Comment by guo-shaoge** [comment]
> tryEliminateProjection

You mean we do not eliminate `Projection` whose output column is empty?

Maybe we can not do that because this [PR](https://github.com/pingcap/tidb/pull/6318)

**Review by lzmhhh123 — COMMENTED** [review]
> > tryEliminateProjection
> 
> You mean we do not eliminate `Projection` whose output column is empty?
> 
> Maybe we can not do that because this [PR](https://github.com/pingcap/tidb/pull/6318)

No, we can check whether a join above it.

**Comment by guo-shaoge** [comment]
> > > tryEliminateProjection
> > 
> > 
> > You mean we do not eliminate `Projection` whose output column is empty?
> > Maybe we can not do that because this [PR](https://github.com/pingcap/tidb/pull/6318)
> 
> No, we can check whether a join above it.

I think this bug is not only for Join, for example, `Apply` may also got same error.

    Apply
     |       |
    DS   Projection
             |
            xxx

SQL: 

    drop table if exists t, s;
    create table t(a int, b int);
    insert into t values(1, 2), (3, 4);
    select (2) in (select b from t) from (select t.a < (select t.a from t t1 limit 1) from t) t;

File: planner/core/rule_column_pruning.go [hunk 1]
```diff
@@ -71,6 +72,16 @@ func (p *LogicalProjection) PruneColumns(parentUsedCols []*expression.Column) er
 			p.Exprs = append(p.Exprs[:i], p.Exprs[i+1:]...)
 		}
 	}
+	// Here we add a constant 1 to avoid Projection operator is eliminated.(#23887)
+	if len(p.Exprs) == 0 {
+		p.schema.Append(&expression.Column{
+			RetType: types.NewFieldType(mysql.TypeLong),
```
Comment (winoros) [inline planner/core/rule_column_pruning.go:78]: Allocate an id for it. `AllocPlanColumnID`

File: planner/core/rule_column_pruning.go [hunk 2]
```diff
@@ -71,6 +72,16 @@ func (p *LogicalProjection) PruneColumns(parentUsedCols []*expression.Column) er
...
+		})
+		p.Exprs = append(p.Exprs, &expression.Constant{
```
Comment (winoros) [inline planner/core/rule_column_pruning.go:80]: We can call `expression.NewNull` Then The column directly using the field type of the expression. No need to create two field type

File: planner/core/rule_column_pruning.go (hunk 1 above)
Comment (guo-shaoge) [reply planner/core/rule_column_pruning.go]: done

File: planner/core/rule_column_pruning.go (hunk 2 above)
Comment (guo-shaoge) [reply planner/core/rule_column_pruning.go:81]: done

**Review by winoros — APPROVED** [review]

**Review by lzmhhh123 — APPROVED** [review]

**Pull request merged**